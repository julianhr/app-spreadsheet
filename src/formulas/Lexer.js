import Token from './Token'

const TOKENS = {
  // internal
  EOF: 'EOF',
  UNKNOWN: 'UNKNOWN',
  // structure
  EQUALS: 'EQUALS',
  COMMA: 'COMMA',
  COLON: 'COLON',
  LPAREN: 'LPAREN',
  RPAREN: 'RPAREN',
  // operators
  PLUS: 'PLUS',
  MINUS: 'MINUS',
  MULT: 'MULT',
  DIV: 'DIV',
  // entities
  NUMBER: 'NUMBER',
  CELL: 'CELL',
  FUNCTION: 'FUNCTION',
}

const t = TOKENS

class Rule {
  constructor(regex, groupIndex, token) {
    this.regex = regex
    this.groupIndex = groupIndex
    this.token = token
  }

  setIndex(index) {
    this.regex.lastIndex = index
  }

  test(input) {
    return this.regex.exec(input)
  }
}

const GRAMMAR = [
  // internal
  new Rule(/=/g, 0, t.EQUALS),
  new Rule(/,/g, 0, t.COMMA),
  new Rule(/:/g, 0, t.COLON),
  new Rule(/\(/g, 0, t.LPAREN),
  new Rule(/\)/g, 0, t.RPAREN),
  // operators
  new Rule(/\+/g, 0, t.PLUS),
  new Rule(/\-/g, 0, t.MINUS), // eslint-disable-line
  new Rule(/\*/g, 0, t.MULT),
  new Rule(/\//g, 0, t.DIV),
  // multi-character
  new Rule(/[\d\.]+/g, 0, t.NUMBER), // eslint-disable-line
  new Rule(/[a-z]+[\d]+/gi, 0, t.CELL),
  new Rule(/([a-z]+)\(/gi, 1, t.FUNCTION),
]


class Lexer {
  constructor(input, grammar) {
    this.input = input
    this.grammar = grammar
    this.index = 0
    this.char = input[this.index]
    this.markers = []
    this.tokens = []
  }

  getTokens() {
    while (this.char !== t.EOF) {
      const token = this.nextToken()
      this.tokens.push(token)
    }

    return this.tokens
  }

  nextToken() {
    if (this.char === t.EOF) {
      return new Token(t.EOF, t.EOF)
    }

    // console.log('before whitespace', this.char, this.index)
    const whitespace = this.getWhitespace()
    let token

    for (let rule of this.grammar) {
      rule.setIndex(this.index)
      const match = rule.test(this.input)

      if (match && match.index === this.index) {
        const text = match[rule.groupIndex]
        token = new Token(rule.token, text, whitespace)
        this.index += text.length - 1
        this.consume()
        break
      }
    }

    if (!token) {
      token = this.tokenUNKNOWN(whitespace)
    }

    return token
  }

  tokenUNKNOWN(whitespace) {
    const chars = []

    while (!(this.isEOF() || this.isSeparator() || this.isWhitespace())) {
      chars.push(this.char)
      this.consume()
    }
    
    const word = chars.join('')
    return new Token(t.UNKNOWN, word, whitespace)
  }

  consume() {
    this.index++

    if (this.index < this.input.length) {
      this.char = this.input[this.index]
    } else {
      this.char = t.EOF
    }
  }

  isEOF() {
    return this.char === t.EOF
  }

  isWhitespace() {
    return !this.isEOF() && Boolean(this.char.match(/\s/))
  }

  isSeparator() {
    return !this.isEOF() && Boolean(this.char.match(/[()+-/*]{1}/))
  }

  getWhitespace() {
    let count = 0

    while (this.isWhitespace()) {
      count++
      this.consume()
    }

    return Array(count).fill(' ').join('')
  }
}

export default Lexer
export { Rule, GRAMMAR, TOKENS, Lexer }
