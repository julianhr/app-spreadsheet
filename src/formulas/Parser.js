import { TOKENS as t } from './Lexer'
import {
  TextNode,
  NumberNode,
  CellNode,
  CellRange,
  BinaryOp,
  UnaryOp,
  FuncOp
} from './ast'


class Parser {
  constructor(tokens) {
    this.index = 0
    this.tokens = tokens
    this.curr = this.tokens[0]
    this.ast = null
  }

  parse() {
    switch (this.peekType()) {
      case t.TEXT:
        this.ast = this.text()
        break
      case t.NUMBER:
        this.ast = this.number()
        break
      default:
        this.equals()
        this.ast = this.expr()
        break
    }

    if (this.curr !== null) {
      throw new Error(`Unexpected term at index ${this.index}: "${this.curr.text}"`)
    }

    return this.ast
  }

  expr() {
    // expr : term ( ( PLUS | MINUS ) term ) *
    let node, operator, right

    node = this.term()

    while ( [t.PLUS, t.MINUS].includes(this.peekType()) ) {
      operator = this.operator()
      right = this.term()
      node = new BinaryOp(node, operator, right)
    }

    return node
  }

  term() {
    // term : factor ( ( MULT | DIV ) factor ) *
    let node, operator, right

    node = this.factor()

    while ( [t.MULT, t.DIV].includes(this.peekType()) ) {
      operator = this.operator()
      right = this.factor()
      node = new BinaryOp(node, operator, right)
    }

    return node
  }

  factor() {
    // factor : ( PLUS | MINUS ) ( NUMBER | CELL )
    //          | LPAREN expr RPAREN
    //          | FUNCTION LPAREN list RPAREN
    switch (this.peekType()) {
      case t.PLUS:
      case t.MINUS:
        return this.unaryOp()
      case t.NUMBER:
        return this.number()
      case t.CELL:
        return this.cell()
      case t.LPAREN:
        return this.enclosedExpr()
      case t.FUNCTION:
        return this.func()
      case t.RPAREN:
        throw new Error(`Unexpected term at index ${this.index}: "${this.curr.text}"`)
      default:
        throw new Error('Missing factor')
    }
  }

  text() {
    let node

    if (this.peekType() !== t.TEXT) {
      throw new Error('Missing string')
    }

    node = new TextNode(this.curr)
    this.consume()
    return node
  }

  number() {
    if (this.peekType() === t.NUMBER) {
      const node = new NumberNode(this.curr)
      this.consume()
      return node
    } else {
      throw new Error('Missing number')
    }
  }

  operator() {
    if ([t.PLUS, t.MINUS, t.MULT, t.DIV].includes(this.peekType())) {
      const node = this.curr
      this.consume()
      return node
    } else {
      throw new Error('Missing operator')
    }
  }

  enclosedExpr() {
    this.lparen()
    const node = this.expr()
    this.rparen()
    return node
  }

  cell() {
    if (this.peekType() !== t.CELL) {
      throw new Error('Missing cell node')
    }

    const node = new CellNode(this.curr)
    this.consume()
    return node
  }

  isCellRange() {
    return (
      this.peekType(0) === t.CELL
      && this.peekType(1) === t.COLON
      && this.peekType(2) === t.CELL
    )
  }

  cellRange() {
    let left, right

    if (!this.isCellRange()) {
      throw new Error('Missing cell range')
    }

    left = new CellNode(this.curr)
    this.consume() // skip colon
    this.consume()
    right = new CellNode(this.curr)
    this.consume()

    if (right.location < left.location) {
      [left, right] = this.invertCellRange(left, right)
    }

    return new CellRange(left, right)
  }

  invertCellRange(left, right) {
    this.tokens[this.index - 2] = right.cell
    this.tokens[this.index] = left.cell
    let temp = left
    left = right
    right = temp
    return [left, right]
  }

  func() {
    if (this.peekType() !== t.FUNCTION) {
      throw new Error('Missing function')
    }

    const func = this.curr
    this.consume()
    this.lparen()
    const args = this.list()
    this.rparen()
    return new FuncOp(func, args)
  }

  list() {
    const list = []
    let node

    node = this.getListNode()
    list.push(node)

    while (this.peekType() === t.COMMA) {
      this.consume() // skip comma token
      node = this.getListNode()
      list.push(node)
    }

    return list
  }

  getListNode() {
    if (this.isCellRange()) {
      return this.cellRange()
    } else {
      return this.expr()
    }
  }

  equals() {
    if (this.peekType() !== t.EQUALS) {
      throw new Error('Missing equals sign')
    }

    this.consume()
  }

  lparen() {
    if (this.peekType() !== t.LPAREN) {
      throw new Error('Missing left parenthesis')
    }

    this.consume()
  }

  rparen() {
    if (this.peekType() !== t.RPAREN) {
      throw new Error('Missing right parenthesis')
    }

    this.consume()
  }

  unaryOp() {
    if (![t.PLUS, t.MINUS].includes(this.peekType())) {
      throw new Error('Missing unary operator')
    }

    const curr = this.curr
    this.consume()
    const node = new UnaryOp(curr, this.factor())
    return node
  }

  consume() {
    if (this.index < this.tokens.length - 1) {
      this.index++
      this.curr = this.tokens[this.index]
    } else {
      this.curr = null
    }
  }

  peekType(stepsAhead) {
    if (stepsAhead) {
      if (stepsAhead < 0) {
        throw new Error('peak ahead index must be positive')
      }

      const index = this.index + stepsAhead

      if (index < this.tokens.length) {
        return this.tokens[index].type
      }
    } else {
      return this.curr && this.curr.type
    }
  }
}

export default Parser
