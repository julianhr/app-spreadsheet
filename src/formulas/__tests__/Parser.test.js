import { Lexer, GRAMMAR } from '../Lexer'
import Parser from '../Parser'


const toThrowWithMsg = {
  toThrowWithMsg: (received) => {
    try {
      received.func()
      return {
        pass: false,
        message: () => `received input "${received.input}" did not throw`
      }
    } catch (error) {
      return {
        pass: true,
        message: () => `received input "${received.input}" threw error`
      }
    }
  }
}

function getParser(input) {
  const lexer = new Lexer(input, GRAMMAR)
  const tokens = lexer.getTokens()
  return new Parser(tokens)
}

describe('Parser', () => {
  describe('#parse', () => {
    describe('happy path', () => {
      test('"=2"', () => {
        const input = '=2'
        const parser = getParser(input)
        const ast = parser.parse()
        expect(ast).toMatchSnapshot()
      })
  
      test('"=2+3"', () => {
        const input = '=2+3'
        const parser = getParser(input)
        const ast = parser.parse()
        expect(ast).toMatchSnapshot()
      })
  
      test('"=2+3/10+8"', () => {
        const input = '=2+3/10+8'
        const parser = getParser(input)
        const ast = parser.parse()
        expect(ast).toMatchSnapshot()
      })
  
      test('"=(2+3)*5"', () => {
        const input = '=(2+3)*5'
        const parser = getParser(input)
        const ast = parser.parse()
        expect(ast).toMatchSnapshot()
      })
  
      test('"=sum(5)"', () => {
        const input = '=sum(5)'
        const parser = getParser(input)
        const ast = parser.parse()
        expect(ast).toMatchSnapshot()
      })
  
      test('"=sum(5,6,7)"', () => {
        const input = '=sum(5,6,7)'
        const parser = getParser(input)
        const ast = parser.parse()
        expect(ast).toMatchSnapshot()
      })
  
      test('"=sum(5, SUM(7,6) ,  (10+5/7))"', () => {
        const input = '=sum(5, SUM(7,6) ,  (10+5/7))'
        const parser = getParser(input)
        const ast = parser.parse()
        expect(ast).toMatchSnapshot()
      })
    })

    describe('error path', () => {
      test('=2+', () => {
        const input = '=2+'
        const parser = getParser(input)
        expect(() => parser.parse()).toThrow('Missing factor')
      })

      test('=(2', () => {
        const input = '=(2'
        const parser = getParser(input)
        expect(() => parser.parse()).toThrow('Missing right parenthesis')
      })

      test('=sum((2,3)', () => {
        const input = '=sum((2,3)'
        const parser = getParser(input)
        expect(() => parser.parse()).toThrow('Missing right parenthesis')
      })

      test('=((2*3)', () => {
        const input = '=((2*3)'
        const parser = getParser(input)
        expect(() => parser.parse()).toThrow('Missing right parenthesis')
      })

      test('=2)', () => {
        const input = '=2)'
        const parser = getParser(input)
        expect(() => parser.parse()).toThrow(/Unexpected term/)
      })

      test('=2)+3', () => {
        const input = '=2)+3'
        const parser = getParser(input)
        expect(() => parser.parse()).toThrow(/Unexpected term/)
      })

      test('=2+)3', () => {
        const input = '=2)+3'
        const parser = getParser(input)
        expect(() => parser.parse()).toThrow(/Unexpected term/)
      })

      test('=(2+3))', () => {
        const input = '=(2+3))'
        const parser = getParser(input)
        expect(() => parser.parse()).toThrow(/Unexpected term/)
      })
    })
  })
})
