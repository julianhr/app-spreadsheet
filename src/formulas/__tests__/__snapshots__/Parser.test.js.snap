// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser #parse happy path arithmitic "=(2+3)*5" 1`] = `
BinaryOp {
  "_name": "BinaryOp",
  "leftNode": BinaryOp {
    "_name": "BinaryOp",
    "leftNode": NumberNode {
      "_name": "NumberNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 2,
        "text": "2",
        "type": "NUMBER",
        "value": 2,
        "whitespace": 0,
      },
    },
    "opNode": Token {
      "category": "operator",
      "html": null,
      "index": 3,
      "text": "+",
      "type": "PLUS",
      "value": null,
      "whitespace": 0,
    },
    "rightNode": NumberNode {
      "_name": "NumberNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 4,
        "text": "3",
        "type": "NUMBER",
        "value": 3,
        "whitespace": 0,
      },
    },
  },
  "opNode": Token {
    "category": "operator",
    "html": null,
    "index": 6,
    "text": "*",
    "type": "MULT",
    "value": null,
    "whitespace": 0,
  },
  "rightNode": NumberNode {
    "_name": "NumberNode",
    "node": Token {
      "category": "entity",
      "html": null,
      "index": 7,
      "text": "5",
      "type": "NUMBER",
      "value": 5,
      "whitespace": 0,
    },
  },
}
`;

exports[`Parser #parse happy path arithmitic "=2" 1`] = `
NumberNode {
  "_name": "NumberNode",
  "node": Token {
    "category": "entity",
    "html": null,
    "index": 1,
    "text": "2",
    "type": "NUMBER",
    "value": 2,
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path arithmitic "=2+-3" 1`] = `
BinaryOp {
  "_name": "BinaryOp",
  "leftNode": NumberNode {
    "_name": "NumberNode",
    "node": Token {
      "category": "entity",
      "html": null,
      "index": 1,
      "text": "2",
      "type": "NUMBER",
      "value": 2,
      "whitespace": 0,
    },
  },
  "opNode": Token {
    "category": "operator",
    "html": null,
    "index": 2,
    "text": "+",
    "type": "PLUS",
    "value": null,
    "whitespace": 0,
  },
  "rightNode": UnaryOp {
    "_name": "UnaryOp",
    "expr": NumberNode {
      "_name": "NumberNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 4,
        "text": "3",
        "type": "NUMBER",
        "value": 3,
        "whitespace": 0,
      },
    },
    "opNode": Token {
      "category": "operator",
      "html": null,
      "index": 3,
      "text": "-",
      "type": "MINUS",
      "value": null,
      "whitespace": 0,
    },
  },
}
`;

exports[`Parser #parse happy path arithmitic "=2+3" 1`] = `
BinaryOp {
  "_name": "BinaryOp",
  "leftNode": NumberNode {
    "_name": "NumberNode",
    "node": Token {
      "category": "entity",
      "html": null,
      "index": 1,
      "text": "2",
      "type": "NUMBER",
      "value": 2,
      "whitespace": 0,
    },
  },
  "opNode": Token {
    "category": "operator",
    "html": null,
    "index": 2,
    "text": "+",
    "type": "PLUS",
    "value": null,
    "whitespace": 0,
  },
  "rightNode": NumberNode {
    "_name": "NumberNode",
    "node": Token {
      "category": "entity",
      "html": null,
      "index": 3,
      "text": "3",
      "type": "NUMBER",
      "value": 3,
      "whitespace": 0,
    },
  },
}
`;

exports[`Parser #parse happy path arithmitic "=2+3/10+8" 1`] = `
BinaryOp {
  "_name": "BinaryOp",
  "leftNode": BinaryOp {
    "_name": "BinaryOp",
    "leftNode": NumberNode {
      "_name": "NumberNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 1,
        "text": "2",
        "type": "NUMBER",
        "value": 2,
        "whitespace": 0,
      },
    },
    "opNode": Token {
      "category": "operator",
      "html": null,
      "index": 2,
      "text": "+",
      "type": "PLUS",
      "value": null,
      "whitespace": 0,
    },
    "rightNode": BinaryOp {
      "_name": "BinaryOp",
      "leftNode": NumberNode {
        "_name": "NumberNode",
        "node": Token {
          "category": "entity",
          "html": null,
          "index": 3,
          "text": "3",
          "type": "NUMBER",
          "value": 3,
          "whitespace": 0,
        },
      },
      "opNode": Token {
        "category": "operator",
        "html": null,
        "index": 4,
        "text": "/",
        "type": "DIV",
        "value": null,
        "whitespace": 0,
      },
      "rightNode": NumberNode {
        "_name": "NumberNode",
        "node": Token {
          "category": "entity",
          "html": null,
          "index": 5,
          "text": "10",
          "type": "NUMBER",
          "value": 10,
          "whitespace": 0,
        },
      },
    },
  },
  "opNode": Token {
    "category": "operator",
    "html": null,
    "index": 7,
    "text": "+",
    "type": "PLUS",
    "value": null,
    "whitespace": 0,
  },
  "rightNode": NumberNode {
    "_name": "NumberNode",
    "node": Token {
      "category": "entity",
      "html": null,
      "index": 8,
      "text": "8",
      "type": "NUMBER",
      "value": 8,
      "whitespace": 0,
    },
  },
}
`;

exports[`Parser #parse happy path cells "= A1 +5" 1`] = `
BinaryOp {
  "_name": "BinaryOp",
  "leftNode": CellNode {
    "_name": "CellNode",
    "cell": Token {
      "category": "entity",
      "html": null,
      "index": 2,
      "text": "A1",
      "type": "CELL",
      "value": "A-1",
      "whitespace": 1,
    },
  },
  "opNode": Token {
    "category": "operator",
    "html": null,
    "index": 5,
    "text": "+",
    "type": "PLUS",
    "value": null,
    "whitespace": 1,
  },
  "rightNode": NumberNode {
    "_name": "NumberNode",
    "node": Token {
      "category": "entity",
      "html": null,
      "index": 6,
      "text": "5",
      "type": "NUMBER",
      "value": 5,
      "whitespace": 0,
    },
  },
}
`;

exports[`Parser #parse happy path cells "=A1" 1`] = `
CellNode {
  "_name": "CellNode",
  "cell": Token {
    "category": "entity",
    "html": null,
    "index": 1,
    "text": "A1",
    "type": "CELL",
    "value": "A-1",
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path cells "=A1+ B2" 1`] = `
BinaryOp {
  "_name": "BinaryOp",
  "leftNode": CellNode {
    "_name": "CellNode",
    "cell": Token {
      "category": "entity",
      "html": null,
      "index": 1,
      "text": "A1",
      "type": "CELL",
      "value": "A-1",
      "whitespace": 0,
    },
  },
  "opNode": Token {
    "category": "operator",
    "html": null,
    "index": 3,
    "text": "+",
    "type": "PLUS",
    "value": null,
    "whitespace": 0,
  },
  "rightNode": CellNode {
    "_name": "CellNode",
    "cell": Token {
      "category": "entity",
      "html": null,
      "index": 5,
      "text": "B2",
      "type": "CELL",
      "value": "B-2",
      "whitespace": 1,
    },
  },
}
`;

exports[`Parser #parse happy path formulas "=sum(5)" 1`] = `
FuncOp {
  "_name": "FuncOp",
  "argNodes": Array [
    NumberNode {
      "_name": "NumberNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 5,
        "text": "5",
        "type": "NUMBER",
        "value": 5,
        "whitespace": 0,
      },
    },
  ],
  "funcNode": Token {
    "category": "entity",
    "html": null,
    "index": 1,
    "text": "sum",
    "type": "FUNCTION",
    "value": [Function],
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path formulas "=sum(5, SUM(7,6) ,  (10+5/7))" 1`] = `
FuncOp {
  "_name": "FuncOp",
  "argNodes": Array [
    NumberNode {
      "_name": "NumberNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 5,
        "text": "5",
        "type": "NUMBER",
        "value": 5,
        "whitespace": 0,
      },
    },
    FuncOp {
      "_name": "FuncOp",
      "argNodes": Array [
        NumberNode {
          "_name": "NumberNode",
          "node": Token {
            "category": "entity",
            "html": null,
            "index": 12,
            "text": "7",
            "type": "NUMBER",
            "value": 7,
            "whitespace": 0,
          },
        },
        NumberNode {
          "_name": "NumberNode",
          "node": Token {
            "category": "entity",
            "html": null,
            "index": 14,
            "text": "6",
            "type": "NUMBER",
            "value": 6,
            "whitespace": 0,
          },
        },
      ],
      "funcNode": Token {
        "category": "entity",
        "html": null,
        "index": 8,
        "text": "SUM",
        "type": "FUNCTION",
        "value": [Function],
        "whitespace": 1,
      },
    },
    BinaryOp {
      "_name": "BinaryOp",
      "leftNode": NumberNode {
        "_name": "NumberNode",
        "node": Token {
          "category": "entity",
          "html": null,
          "index": 21,
          "text": "10",
          "type": "NUMBER",
          "value": 10,
          "whitespace": 0,
        },
      },
      "opNode": Token {
        "category": "operator",
        "html": null,
        "index": 23,
        "text": "+",
        "type": "PLUS",
        "value": null,
        "whitespace": 0,
      },
      "rightNode": BinaryOp {
        "_name": "BinaryOp",
        "leftNode": NumberNode {
          "_name": "NumberNode",
          "node": Token {
            "category": "entity",
            "html": null,
            "index": 24,
            "text": "5",
            "type": "NUMBER",
            "value": 5,
            "whitespace": 0,
          },
        },
        "opNode": Token {
          "category": "operator",
          "html": null,
          "index": 25,
          "text": "/",
          "type": "DIV",
          "value": null,
          "whitespace": 0,
        },
        "rightNode": NumberNode {
          "_name": "NumberNode",
          "node": Token {
            "category": "entity",
            "html": null,
            "index": 26,
            "text": "7",
            "type": "NUMBER",
            "value": 7,
            "whitespace": 0,
          },
        },
      },
    },
  ],
  "funcNode": Token {
    "category": "entity",
    "html": null,
    "index": 1,
    "text": "sum",
    "type": "FUNCTION",
    "value": [Function],
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path formulas "=sum(5,6)" 1`] = `
FuncOp {
  "_name": "FuncOp",
  "argNodes": Array [
    NumberNode {
      "_name": "NumberNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 5,
        "text": "5",
        "type": "NUMBER",
        "value": 5,
        "whitespace": 0,
      },
    },
    NumberNode {
      "_name": "NumberNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 7,
        "text": "6",
        "type": "NUMBER",
        "value": 6,
        "whitespace": 0,
      },
    },
    NumberNode {
      "_name": "NumberNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 9,
        "text": "7",
        "type": "NUMBER",
        "value": 7,
        "whitespace": 0,
      },
    },
  ],
  "funcNode": Token {
    "category": "entity",
    "html": null,
    "index": 1,
    "text": "sum",
    "type": "FUNCTION",
    "value": [Function],
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path formulas "=sum(A1:B2, C1)" 1`] = `
FuncOp {
  "_name": "FuncOp",
  "argNodes": Array [
    CellRange {
      "_name": "CellRange",
      "leftCell": CellNode {
        "_name": "CellNode",
        "cell": Token {
          "category": "entity",
          "html": null,
          "index": 5,
          "text": "A1",
          "type": "CELL",
          "value": "A-1",
          "whitespace": 0,
        },
      },
      "rightCell": CellNode {
        "_name": "CellNode",
        "cell": Token {
          "category": "entity",
          "html": null,
          "index": 8,
          "text": "B2",
          "type": "CELL",
          "value": "B-2",
          "whitespace": 0,
        },
      },
    },
    CellNode {
      "_name": "CellNode",
      "cell": Token {
        "category": "entity",
        "html": null,
        "index": 12,
        "text": "C1",
        "type": "CELL",
        "value": "C-1",
        "whitespace": 1,
      },
    },
  ],
  "funcNode": Token {
    "category": "entity",
    "html": null,
    "index": 1,
    "text": "sum",
    "type": "FUNCTION",
    "value": [Function],
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path formulas "=sum(B2:A1, C2:D4)" 1`] = `
FuncOp {
  "_name": "FuncOp",
  "argNodes": Array [
    CellRange {
      "_name": "CellRange",
      "leftCell": CellNode {
        "_name": "CellNode",
        "cell": Token {
          "category": "entity",
          "html": null,
          "index": 8,
          "text": "A1",
          "type": "CELL",
          "value": "A-1",
          "whitespace": 0,
        },
      },
      "rightCell": CellNode {
        "_name": "CellNode",
        "cell": Token {
          "category": "entity",
          "html": null,
          "index": 5,
          "text": "B2",
          "type": "CELL",
          "value": "B-2",
          "whitespace": 0,
        },
      },
    },
    CellRange {
      "_name": "CellRange",
      "leftCell": CellNode {
        "_name": "CellNode",
        "cell": Token {
          "category": "entity",
          "html": null,
          "index": 12,
          "text": "C2",
          "type": "CELL",
          "value": "C-2",
          "whitespace": 1,
        },
      },
      "rightCell": CellNode {
        "_name": "CellNode",
        "cell": Token {
          "category": "entity",
          "html": null,
          "index": 15,
          "text": "D4",
          "type": "CELL",
          "value": "D-4",
          "whitespace": 0,
        },
      },
    },
  ],
  "funcNode": Token {
    "category": "entity",
    "html": null,
    "index": 1,
    "text": "sum",
    "type": "FUNCTION",
    "value": [Function],
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path formulas '=CONCAT("str1"," str2 ")' 1`] = `
FuncOp {
  "_name": "FuncOp",
  "argNodes": Array [
    StringNode {
      "_name": "StringNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 8,
        "text": "\\"str1\\"",
        "type": "STRING",
        "value": "str1",
        "whitespace": 0,
      },
    },
    StringNode {
      "_name": "StringNode",
      "node": Token {
        "category": "entity",
        "html": null,
        "index": 15,
        "text": "\\" str2 \\"",
        "type": "STRING",
        "value": " str2 ",
        "whitespace": 0,
      },
    },
  ],
  "funcNode": Token {
    "category": "entity",
    "html": null,
    "index": 1,
    "text": "CONCAT",
    "type": "FUNCTION",
    "value": [Function],
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path strings "  2" 1`] = `
TextNode {
  "_name": "TextNode",
  "node": Token {
    "category": "entity",
    "html": null,
    "index": 0,
    "text": "  2",
    "type": "TEXT",
    "value": null,
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path strings "  2+" 1`] = `
TextNode {
  "_name": "TextNode",
  "node": Token {
    "category": "entity",
    "html": null,
    "index": 0,
    "text": "  2+",
    "type": "TEXT",
    "value": null,
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path strings "  2+3" 1`] = `
TextNode {
  "_name": "TextNode",
  "node": Token {
    "category": "entity",
    "html": null,
    "index": 0,
    "text": "  2+3",
    "type": "TEXT",
    "value": null,
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path strings "  text" 1`] = `
TextNode {
  "_name": "TextNode",
  "node": Token {
    "category": "entity",
    "html": null,
    "index": 0,
    "text": "  text",
    "type": "TEXT",
    "value": null,
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path strings " " 1`] = `
TextNode {
  "_name": "TextNode",
  "node": Token {
    "category": "entity",
    "html": null,
    "index": 0,
    "text": " ",
    "type": "TEXT",
    "value": null,
    "whitespace": 0,
  },
}
`;

exports[`Parser #parse happy path strings "" 1`] = `
NumberNode {
  "_name": "NumberNode",
  "node": Token {
    "category": "entity",
    "html": null,
    "index": -1,
    "text": "0",
    "type": "NUMBER",
    "value": 0,
    "whitespace": 0,
  },
}
`;
